File: mid/19.cpp
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        ListNode *first = head, *second = head;
        for(int i = 0; i < n; i++) { // first向前走n步
            first = first->next;
        }
        if(first == NULL) { // first走到了NULL，要删除的是第一个结点
            ListNode *toDelete = second;
            second = second->next;
            delete toDelete;
            toDelete = NULL;
            return second;

        }
        while(first->next != NULL) { // first还没走到NULL
            first = first->next;
            second = second->next;
        }
        ListNode *toDelete = second->next; // second->toDelete->node
        second->next = second->next->next;
        delete toDelete;
        toDelete = NULL;

        return head;
    }
};

File: mid/19.py
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        listNode1 = head
        for i in range(n): # listNode1向前走 n 步
            listNode1 = listNode1.next

        if listNode1 == None: # listNode1走到了NULL，要删除的是第一个结点
            toDelete = head
            head = head.next
            del toDelete
            return head
        
        listNode2 = head
        while listNode1.next != None: # listNode2->toDelete->node
            listNode1 = listNode1.next
            listNode2 = listNode2.next
        
        toDelete = listNode2.next
        listNode2.next = listNode2.next.next
        del toDelete
        return head